test/e2e/magic_catalog.go

```diff
@@ -78,46 +75,93 @@ func (c *magicCatalog) DeployCatalog(ctx context.Context) error {
 		}
 
 		state := catalogSource.Status.GRPCConnectionState.LastObservedState
-
 		if state != catalogReadyState {
 			return false, nil
-		} else {
-			return true, nil
 		}
+		return true, nil
 	})
-
 	if err != nil {
 		return c.cleanUpAfter(ctx, err)
 	}
 
 	return nil
 }
 
-func (c *magicCatalog) UndeployCatalog(ctx context.Context) []error {
-	var errs []error = nil
+func (c *magicCatalog) deployCatalog(ctx context.Context, resources []k8scontrollerclient.Object) error {
+	for _, res := range resources {
+		err := c.kubeClient.Create(ctx, res)
+		if err != nil {
+			return c.cleanUpAfter(ctx, err)
+		}
+	}
+	return nil
+}
+
+func (c *magicCatalog) UpdateCatalog(ctx context.Context, provider FileBasedCatalogProvider) error {
+	resourcesInOrderOfDeletion := []k8scontrollerclient.Object{
+		c.makeCatalogSourcePod(),
+		c.makeConfigMap(),
+	}
+	errors := c.undeployCatalog(ctx, resourcesInOrderOfDeletion)
+	if len(errors) != 0 {
+		return utilerrors.NewAggregate(errors)
+	}
+
+	// TODO(tflannag): Create a pod watcher struct and setup an underlying watch
+	// and block until ctx.Done()?
+	err := waitFor(func() (bool, error) {
+		pod := &corev1.Pod{}
+		err := c.kubeClient.Get(ctx, k8scontrollerclient.ObjectKey{
+			Name:      c.podName,
+			Namespace: c.namespace,
+		}, pod)
+		if k8serror.IsNotFound(err) {
+			return true, nil
+		}
+		return false, err
+	})
+	if err != nil {
+		return fmt.Errorf("failed to successfully update the catalog deployment: %v", err)
+	}
 
+	c.fileBasedCatalog = provider
```

Comment:
Note: we only update the underlying provider until after we delete the registry pod and configmap resources attached to the previous provider. This fbc_provider.go package might need a setter method, or use the options pattern so NewMagicCatalog().WithProvider(...) is slightly more readable?